Controller/Adminhtml/Products/SyncData.php

```diff
@@ -0,0 +1,136 @@
+        CategoryFactory   $categoryFactory
+    ) {
+        parent::__construct($context);
+        $this->syncHelper               = $syncHelper;
+        $this->productCollectionFactory = $productCollectionFactory;
+        $this->imageHelper              = $imageHelper;
+        $this->categoryFactory          = $categoryFactory;
+    }
+
+    public function execute()
+    {
+        $productCollection = $this->productCollectionFactory->create();
+        $productCollection->addAttributeToSelect(['name', 'price', 'status', 'sku', 'image', 'description']);
+//        $productCollection->addAttributeToFilter('status', ['eq' => \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED]);
+
+        $data = [];
+        foreach ($productCollection as $product) {
+            if ($product->getTypeId() == \Magento\ConfigurableProduct\Model\Product\Type\Configurable::TYPE_CODE) {
+                // Handle configurable product
+                $childProducts = $product->getTypeInstance()->getUsedProducts($product);
+                foreach ($childProducts as $child) {
+                    $stockItem = $child->getExtensionAttributes()->getStockItem();
+                    $quantity  = $stockItem ? $stockItem->getQty() : 0;
+                    if ($quantity === null) {
+                        $quantity = $child->getIsInStock() ? 1 : 0;
+                    }
+
+                    $data[] = [
+                        'name'         => $product->getName(),
+                        'product_id'   => $product->getId(),
+                        'variant_id'   => $child->getId(),
+                        'variant_name' => $child->getName(),
+                        'quantity'     => $quantity,
+                        'image'        => $this->imageHelper->init($child, 'product_page_image_small')->getUrl(),
+                        'is_published' => $product->getStatus() == \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED,
+                        'price'        => $child->getPrice(),
+                        'old_price'    => $child->getPrice() !== $child->getSpecialPrice() ? $child->getSpecialPrice() : 0,
+                        'description'  => $product->getDescription(),
+                        'sku'          => $child->getSku(),
+                        'sub_category' => $this->getProductCategory($product),
+                        'category'     => $this->getProductCategory($product, true),
+                    ];
+                }
+            } elseif ($product->getTypeId() == \Magento\Catalog\Model\Product\Type::TYPE_SIMPLE) {
+                // Handle simple product
+                $stockItem = $product->getExtensionAttributes()->getStockItem();
+                $quantity  = $stockItem ? $stockItem->getQty() : 0;
+                if ($quantity === null) {
+                    $quantity = $product->getIsInStock() ? 1 : 0;
+                }
+
+                $data[] = [
```

Comment:
The `data` array here is almost the same as the one on line `51`, with variant_id and variant_name being the only distinction. Can we build this array at the top and only bring in those variants if they exist?

Reply:
That'll require declaring a variable at line 41 and then populate it and finally adding it to the data array after line 90.